File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intvl: List[List[int]], new: List[int]) -> List[List[int]]:
        intvl.append(new)
        intvl.sort()
        st, end = intvl[0]
        ans=[]
        for i in range(1, len(intvl)):
            if intvl[i][0]<=end:
                end = max(end, intvl[i][1])
            else:
                ans.append([st, end])
                st, end = intvl[i][0], intvl[i][1]
        ans.append([st, end])
        return ans

File: 0057-insert-interval/0057-insert-interval.java
// O(logn + n)

class Solution {
    public int[][] insert(int[][] intvl, int[] newIntvl) {
        List<int[]> ans = new ArrayList<>();
        int idx = bs(intvl, newIntvl[0]);
        
        for(int i=0; i<idx; ++i) ans.add(intvl[i]);
        int i = ans.size();
        
        if(i!=0 && ans.get(i-1)[1] >= newIntvl[0])
            ans.get(i-1)[1] = Math.max(newIntvl[1], ans.get(i-1)[1]);
        else
            ans.add(newIntvl);
        
        for(; i<intvl.length; ++i){
            if(ans.get(ans.size()-1)[1] >= intvl[i][0])
                ans.get(ans.size()-1)[1] = Math.max(ans.get(ans.size()-1)[1], intvl[i][1]);
            else
                ans.add(intvl[i]);
        }
        int[][] res = new int[ans.size()][2];
        for(int j=0; j<ans.size(); ++j) res[j] = ans.get(j);
        return res;
    }
    
    int bs(int[][] arr, int x){
        int i=0;
        int j=arr.length;
        while(i<j){
            int mid = i+(j-i)/2;
            if(arr[mid][0] < x) i=mid + 1;
            else j = mid;
        }
        return j;
    }
}